internal/common/awsxray/tracesegment.go

```diff
@@ -0,0 +1,270 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsxray
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+)
+
+const (
+	// TypeStr is the type and ingest format of this receiver
+	TypeStr = "awsxray"
+)
+
+// Header stores header of trace segment.
+type Header struct {
+	Format  string `json:"format"`
+	Version int    `json:"version"`
+}
+
+// IsValid validates Header.
+func (t Header) IsValid() bool {
+	return strings.EqualFold(t.Format, "json") && t.Version == 1
+}
+
+type CauseType int
+
+const (
+	// CauseTypeExceptionID indicates that the type of the `cause`
+	// field is a string
+	CauseTypeExceptionID CauseType = iota + 1
+	// CauseTypeObject indicates that the type of the `cause`
+	// field is an object
+	CauseTypeObject
+)
+
+// Segment schema is documented in xray-segmentdocument-schema-v1.0.0 listed
+// on https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html
+type Segment struct {
+	// Required fields for both segment and subsegments
+	Name      *string  `json:"name"`
+	ID        *string  `json:"id"`
+	StartTime *float64 `json:"start_time"`
+
+	// Segment-only optional fields
+	Service     *ServiceData `json:"service,omitempty"`
+	Origin      *string      `json:"origin,omitempty"`
```

Comment:
I personally like to distinguish between a default value and a value explicitly provided. For example, if a field is of type `bool` and its value is `false`, I can't really tell whether it is explicitly set to `false` or Go just uses its default value so I still prefer pointer types. Go AWS SDK also follows this pattern. However, I definitely agree that having to write `aws.XXXX` is definitely very annoying, kinda wish Go will give us a better type system.

Reply:
Thanks - I realized that in the exporter the primitive types are fine, but not in the receiver because of that issue. So using pointers here in the common schema makes sense to me, it's indeed conventional was just wondering if it's actually required and it does seem to be.